### PR TITLE
Oleg.vyalyh/fcm

### PR DESCRIPTION
**Pull request by oleg-vyalyh**

### CODE REVIEW

File: src/com/backendless/Messaging.java [hunk 1]
```diff
@@ -217,6 +250,77 @@ private synchronized void registerDeviceGCMSync( Context context, String GCMSend
...
+  }
+
+  /**
+   * For FireBase messaging only.
+   * @throws BackendlessException
+   */
+  public void registerDevice( List<String> channels, Date expiration, AsyncCallback<String> callback ) throws BackendlessException
+  {
+    if( !BackendlessPushService.isFCM( ContextHandler.getAppContext() ) )
+      throw new IllegalStateException( "The method is intended only for FireBase messaging." );
```
Comment (ksv510) [inline src/com/backendless/Messaging.java]: we should improve this error message with information what to do to fix the error

File: src/com/backendless/Messaging.java [hunk 2]
```diff
@@ -217,6 +250,77 @@ private synchronized void registerDeviceGCMSync( Context context, String GCMSend
...
+
+    if( channels == null || channels.isEmpty() ||
+            (channels.size() == 1 && (channels.get( 0 ) == null || channels.get( 0 ).isEmpty())) )
+      channels = Collections.singletonList( DEFAULT_CHANNEL_NAME );
```
Comment (ksv510) [inline src/com/backendless/Messaging.java]: add curly brackets

File: src/com/backendless/Messaging.java [hunk 3]
```diff
@@ -217,6 +250,77 @@ private synchronized void registerDeviceGCMSync( Context context, String GCMSend
...
+
+    for( String channel : channels )
+      checkChannelName( channel );
+
+    long expirationMs = 0;
+    if( expiration != null)
```
Comment (ksv510) [inline src/com/backendless/Messaging.java]: add curly brackets

File: src/com/backendless/Messaging.java [hunk 4]
```diff
@@ -301,10 +415,15 @@ protected RuntimeException doInBackground( Void... params )
         {
           Context context = ContextHandler.getAppContext();
 
-          if( !GCMRegistrar.isRegistered( context ) )
-            return new IllegalArgumentException( ExceptionMessage.DEVICE_NOT_REGISTERED );
+          if ( BackendlessPushService.isFCM( context ) )
```
Comment (ksv510) [inline src/com/backendless/Messaging.java]: add curly brackets

File: src/com/backendless/push/BackendlessPushService.java [hunk 5]
```diff
@@ -59,18 +75,65 @@
...
+        {
+          if( BackendlessPushService.class.isAssignableFrom( clsLoader.loadClass( srvInfo.name ) ) )
+          {
+            BackendlessPushService.pushServiceClassName = srvInfo.name;
+            break;
+          }
+        }
+        catch( ClassNotFoundException e )
+        {
+          Log.e( TAG, "Can not load declared service class.", e );
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: why this exception occurs and why we ignore it?
if we ignore it maybe we should use log level info or warning?

File: src/com/backendless/push/BackendlessPushService.java [hunk 6]
```diff
@@ -358,27 +442,15 @@ private void registerFurther( final Context context, String GCMregistrationId )
...
-          ids = (String) obj[0];
-          PushTemplateHelper.deleteNotificationChannel( context );
-          Map<String,AndroidPushTemplate> templates = (Map<String,AndroidPushTemplate>) obj[1];
+        Map<String, String> channelRegistrations = processRegistrationPayload( context, registrationInfo );
 
-          if( android.os.Build.VERSION.SDK_INT > 25 )
-            for( AndroidPushTemplate templ : templates.values() )
-              PushTemplateHelper.getOrCreateNotificationChannel( context.getApplicationContext(), templ );
+        StringBuilder sb = new StringBuilder();
+        for( String id : channelRegistrations.values() )
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: add empty line before loop and aftrer

File: src/com/backendless/push/BackendlessPushService.java [hunk 7]
```diff
@@ -407,4 +479,191 @@ public void handleFault( BackendlessFault fault )
...
+  private Map<String, String> processRegistrationPayload( final Context context, final String payload )
+  {
+    Object[] obj;
+    try
+    {
+      obj = (Object[]) weborb.util.io.Serializer.fromBytes( payload.getBytes(), weborb.util.io.Serializer.JSON, false );
+    }
+    catch( IOException e )
+    {
+      callback.onError( context, "Could not deserialize server response: " + e.getMessage() );
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: log this error

File: src/com/backendless/push/BackendlessPushService.java [hunk 8]
```diff
@@ -407,4 +479,191 @@ public void handleFault( BackendlessFault fault )
...
+      return null;
+    }
+
+    PushTemplateHelper.deleteNotificationChannel( context );
+    Map<String, AndroidPushTemplate> templates = (Map<String, AndroidPushTemplate>) obj[ 1 ];
+
+    if( android.os.Build.VERSION.SDK_INT > 25 )
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: add curly brackets

File: src/com/backendless/push/BackendlessPushService.java [hunk 9]
```diff
@@ -407,4 +479,191 @@ public void handleFault( BackendlessFault fault )
...
+      for( AndroidPushTemplate templ : templates.values() )
+        PushTemplateHelper.getOrCreateNotificationChannel( context.getApplicationContext(), templ );
+
+    PushTemplateHelper.setPushNotificationTemplates( templates, payload.getBytes() );
+
+    String regs = (String) obj[ 0 ];
+    Map<String, String> channelRegistrations = new HashMap<>();
+    String[] regPairs = regs.split( "," );
+    for( String pair : regPairs )
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: add empty line before and after loop

File: src/com/backendless/push/BackendlessPushService.java [hunk 10]
```diff
@@ -59,18 +75,65 @@
...
+      }
+    }
+    catch( PackageManager.NameNotFoundException e )
+    {
+      Log.e( TAG, "Can not load current app package." );
+      throw new IllegalStateException( "Can not load current app package." );
+    }
+
+    if( BackendlessPushService.pushServiceClassName == null )
+      throw new IllegalStateException( "Can not find 'com.backendless.push.BackendlessPushService' or it inheritor in AndroidManifest.xml." );
```
Comment (markpiller) [inline src/com/backendless/push/BackendlessPushService.java]: A reference to BackendlessPushService is missing in AndroidManifest.xml. Make sure to add <service android:name="com.backendless.push.BackendlessPushService" /> to the manifest file. Alternatively, if you have a custom push service implementation, make sure your class inherits from "com.backendless.push.BackendlessPushService" and is registered in the manifest file as <service android:name="your push service fully qualified class name" />

File: src/com/backendless/push/FCMRegistration.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        if (callback != null)
+          callback.handleFault( new BackendlessFault( "Failed to subscribe in FCM. " + task.getException().getMessage() ) );
+      }
+      else
+        FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener( new OnCompleteListener<InstanceIdResult>()
+        {
+          @Override
+          public void onComplete( @NonNull Task<InstanceIdResult> task )
+          {
+          if( !task.isSuccessful() )
```
Comment (ksv510) [inline src/com/backendless/push/FCMRegistration.java]: add a margin for if block

File: src/com/backendless/push/BackendlessPushService.java [hunk 12]
```diff
@@ -112,46 +175,74 @@ public void onDestroy()
...
+        handleMessage( this, intent );
+        break;
+    }
+  }
+
+  @Override
+  public void onRegistered( Context context, Map<String, String> channelRegistrations )
+  {
+    StringBuilder sb = new StringBuilder();
+    for( String id : channelRegistrations.values() )
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: add empty line before and after loop

File: src/com/backendless/push/BackendlessPushService.java [hunk 13]
```diff
@@ -59,18 +75,67 @@
...
+    }
+    catch( PackageManager.NameNotFoundException e )
+    {
+      Log.e( TAG, "Can not load current app package." );
+      throw new IllegalStateException( "Can not load current app package." );
+    }
+
+    if( BackendlessPushService.pushServiceClassName == null )
+    {
+      throw new IllegalStateException( "A reference to BackendlessPushService is missing in AndroidManifest.xml. Make sure to add <service android:name=\"com.backendless.push.BackendlessPushService\" /> to the manifest file. Alternatively, if you have a custom push service implementation, make sure your class inherits from \"com.backendless.push.BackendlessPushService\" and is registered in the manifest file as <service android:name=\"your push service fully qualified class name\" />" );
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: line to long

File: src/com/backendless/push/BackendlessPushService.java [hunk 14]
```diff
@@ -407,4 +485,196 @@ public void handleFault( BackendlessFault fault )
...
+
+    ClassLoader clsLoader = appContext.getClassLoader();
+
+    try
+    {
+      clsLoader.loadClass( "com.google.firebase.messaging.FirebaseMessagingService" );
+    }
+    catch( ClassNotFoundException e )
+    {
+      Log.i( TAG, "Class FirebaseMessagingService not found. Try to use GCM." );
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: `Try to use GCM` should be changed `GSM will be used`

File: src/com/backendless/push/BackendlessPushService.java [hunk 15]
```diff
@@ -407,4 +485,196 @@ public void handleFault( BackendlessFault fault )
...
+        }
+        catch( ClassNotFoundException e )
+        {
+          Log.e( TAG, "Can not load declared service class.", e );
+        }
+      }
+
+      if( !flag )
+      {
+        Log.i( TAG, "Class FirebaseMessagingService not found. Try to use GCM." );
```
Comment (ksv510) [inline src/com/backendless/push/BackendlessPushService.java]: `Try to use GCM` change to `GCM will be used`

**Pull request merged**